Add V4 Roles

nklaassen opened the pull request:
Issue #5513 

This PR adds V4 roles, which are identical to V3 roles except that they have stricter defaults for their `allow` labels.

Label | RoleV3 Default | RoleV4 Default
----- | --------------- | ---------------
`NodeLabels` | `[{"*": "*"}]` if the user has any logins, else `[]` | `nil`
`AppLabels` | `[{"*": "*"}]` | `nil`
`KubernetesLabels` | `[{"*": "*"}]` | `nil`
`DatabaseLabels` | `[{"*": "*"}]` | `nil`

Defaults are set for all `nil` label sets (and sometimes for empty label sets, see [Downgrading Roles](#downgrading-roles)), not only when the role is first created but also at various other points (e.g. after serializing/deserializing).

## Backward Compatibility
Existing V3 roles will be unaffected. Users can continue to create V3 roles without changes.

Because V4 roles have an identical structure to V3, they are completely compatible *except* that older clients will not recognize the V4 version tag and throw an error.

Because we want V4 roles to be usable on a cluster with teleport instances running a version which does not have RoleV4 support, the teleport gRPC server will downgrade all V4 roles to V3 before sending them to older clients.

### Detecting Client Version
In order for the server to be able to detect the client version, I have added a version tag to the gRPC metadata that will be sent with all requests from the gRPC client - see changes in `api/metadata/metadata.go` and `api/client/client.go`. If no version metadata is received with a gRPC request, the server will assume that the client is running an older teleport version and send the downgraded V3 role (see `downgradeRole` in `lib/auth/grpcserver.go`).

### Downgrading Roles
Two things need to be done to downgrade a V4 role to V3:
1. The version tag needs to be changed to `v3`. Older instances will not recognize a newer version tag, and throw an error.
2. If any of the label sets are currently empty, set a placeholder label so that the older clients will not mistakenly set the V3 defaults.
At first I thought we could just set the labels to an empty list rather than `nil`, but distinguishing an empty vs a `nil` list is always error prone - and because we currently serialize labels with `omitempty` they may be *impossible* to distinguish. For this reason, I set a placeholder label for any empty label set during the downgrade (see `DowngradeRoleToV3` in `lib/services/role.go`, the current placeholder is `{"__teleport_no_labels": <uuid>}`).

### Rolling Back / Downgrading Teleport
If a user creates any V4 roles, they will not be able to downgrade to a teleport version without V4 role support without first deleting all of the V4 roles, or manually changing them to V3.

## RoleV3 Type Rename
As part of this change `RoleV3` and `RoleSpecV3` should probably also be renamed to `RoleV4` and `RoleSpecV4`. To avoid creating too much noise in this diff, I've put this in a following PR at https://github.com/gravitational/teleport/pull/7132

**Related Issues**

#5513

nklaassen committed:
add v4 roles

nklaassen committed:
send client version in grpc metadata

Joerger commented on api/types/role.go:
```diff
@@ -28,6 +28,7 @@ import (
 
 	"github.com/gogo/protobuf/proto"
 	"github.com/gravitational/trace"
+	"github.com/pborman/uuid"
```
Is this package strictly necessary? If possible, try to limit the dependencies added in the api client.

fspmarshall replied:
Having some kind of random value is important.  How about moving the downgrade logic to a standalone function in `services`?  Thats how we've dealt with limiting `api` imports elsewhere.

nklaassen replied:
I've moved this to lib/services/role.go in the latest commit to avoid the uuid dep in api/

nklaassen commented on api/metadata/metadata.go:
```diff
@@ -0,0 +1,78 @@
+const versionKey = "version"
+
+var (
+	versionValue = teleport.Version
```
I've updated this in the latest version

nklaassen commented on lib/services/types.go:
```diff
@@ -496,6 +496,7 @@ const (
 	KindIdentity                  = types.KindIdentity
 	KindState                     = types.KindState
 	KindKubeService               = types.KindKubeService
+	V4                            = types.V4
```
thanks, removed

nklaassen commented on lib/services/role.go:
```diff
@@ -2334,6 +2334,9 @@ func UnmarshalRole(bytes []byte, opts ...MarshalOption) (Role, error) {
 	switch h.Version {
+	case V4:
+		// V4 roles are identical to V3 except for their defaults
+		fallthrough
```
Interesting idea, so this would look more like adding a field to the struct to decide which defaults you get, rather than incrementing the version? I think that could potentially work (with similar "downgrade" steps). I'm curious what others think though

r0mant commented on api/constants/constants.go:
```diff
@@ -134,3 +134,8 @@ const (
+
+const (
+	// TODO(Joerger): change this to generated value
+	Version = "7.0.0-dev"
```
We already have version in `version.go` file. Is this in addition to that? Does this mean we'd need to bump version in 3 places now when doing releases (Makefile, version.go and here)?

nklaassen replied:
`teleport/api/` doesn't depend on `teleport/` which is why I can't really use the value from `version.go`. It sounds like @Joerger is working on something so that `api/` will have access to the generated client version, the suggestion was to copy this here for now and let him fill in the generated value when his changes come in. I'm not sure how quickly those changes are coming or if we should put in something to get the generated value for this PR @Joerger ? fyi I think we want this going out in the 6.2 train

Joerger replied:
Here's the PR for the client version generation - #7157

r0mant commented on api/metadata/metadata.go:
```diff
@@ -0,0 +1,78 @@
+// SetVersion is used only for tests to set the version value that the client
+// will send in the GRPC metadata.
+func SetVersion(version string) {
+	versionValue = version
```
I would consider protecting the global variables by a mutex or ensure atomicity some other way to account for potential concurrent access.

nklaassen replied:
Good point, added these globals as kind of a hack for testing so didn't really think about it.  I've wrapped access to these in a RWMutex in the latest commit.

r0mant commented on api/types/role.go:
```diff
@@ -612,6 +621,37 @@ func (r *RoleV3) CheckAndSetDefaults() error {
+		// downgraded, set a placeholder label
+		const labelKey = "__teleport_no_labels"
+		labelVal := uuid.NewString()
+		if r.Spec.Allow.NodeLabels == nil || len(r.Spec.Allow.NodeLabels) == 0 {
```
Here and below, checking `len(...) == 0` would be sufficient as it covers the "nil" case as well.

nklaassen replied:
good point, updated

r0mant commented on lib/services/role.go:
```diff
@@ -147,7 +147,7 @@ func NewAdminRole() Role {
 	return &RoleV3{
 		Kind:    KindRole,
-		Version: V3,
+		Version: types.V4,
```
This being V4 role now - does nothing break with implicit role not having any allowed labels set?

nklaassen replied:
well so far nothing broke but thanks for pointing this out, the plan for this PR was actually to leave the builtin roles on V3 so I've reverted the change here and in RoleForUser

nklaassen committed:
addressing review comments

move downgrade func to services/ to avoid uuid dep in api/
don't check list for nil, len() will return 0
add comments

nklaassen commented on api/types/role.go:
```diff
@@ -612,6 +621,37 @@ func (r *RoleV3) CheckAndSetDefaults() error {
 	return nil
 }
 
+// DowngradeToV3 converts a V4 role to V3 so that it will be compatible with older instances.
```
good call. I would have preferred to do this immutably, but seems like I'd have to implement a manual clone function (`proto.Clone` panics)

nklaassen commented on api/types/role.go:
```diff
@@ -612,6 +621,37 @@ func (r *RoleV3) CheckAndSetDefaults() error {
+// DELETE IN 8.0.0
+func (r *RoleV3) DowngradeToV3() error {
+	switch r.Version {
+	case V3:
```
yep added a break comment in the latest commit

nklaassen committed:
revert version change to implicit role and RoleForUser

nklaassen committed:
add mutex for globals in metadata.go

smallinsky commented on api/metadata/metadata.go:
```diff
@@ -0,0 +1,85 @@
+
+// SetVersion is used only for tests to set the version value that the client
+// will send in the GRPC metadata.
+func SetVersion(version string) {
+	mu.Lock()
+	defer mu.Unlock()
+	versionValue = version
+}
+
+// DisabledIs disables (or enables) including metadata in GRPC requests. Used in tests.
+func DisabledIs(metadataDisabled bool) {
+	mu.Lock()
+	defer mu.Unlock()
+	disabled = metadataDisabled
+}
+
+func addMetadataToContext(ctx context.Context) context.Context {
+	mu.RLock()
+	defer mu.RUnlock()
+	if disabled {
+		return ctx
+	}
+	return metadata.AppendToOutgoingContext(ctx, versionKey, versionValue)
+}
```
I'm wondering if these test helper functions can be removed. Instead of operating on global state we can just set context ongoing  metadata  in the test like:

```go
ctx := context.Background()
if  tc.versionValue != "" {
    ctx  = metadata.AppendToOutgoingContext(ctx, versionKey, tc.versionValue)
}
gotRoles, err := client.GetRoles(ctx)
...
```

nklaassen replied:
Thanks for this comment, I hadn't considered that the caller can pass their own metadata through the client, I've updated the metadata package and the test to avoid using globals in the latest commit

smallinsky commented on api/metadata/metadata.go:
```diff
@@ -0,0 +1,85 @@
+// VersionFromContext can be called from a GRPC server method to return the
+// client version that was added to the GRPC metadata by
+// StreamClientInterceptor or UnaryClientInterceptor on the client.
+func VersionFromContext(ctx context.Context) (string, bool) {
```
```suggestion
func ClientVersionFromContext(ctx context.Context) (string, bool) {
```

nklaassen replied:
updated in latest commit

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
s/VersionFromContext/ClientVersionFromContext/

nklaassen committed:
avoid mutable globals in metadata package

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
lint fix

fspmarshall commented on lib/services/role.go:
```diff
@@ -2210,3 +2214,34 @@ func MarshalRole(role types.Role, opts ...MarshalOption) ([]byte, error) {
+
+// DowngradeToV3 converts a V4 role to V3 in-place so that it will be compatible with older instances.
+// DELETE IN 8.0.0
+func DowngradeRoleToV3(r *types.RoleV3) error {
```
Nit: Not sure if its *necessary*, but I'd feel better if the downgrade process returned a new value that was a shallow copy in the event that the downgrade needed to occur.  To my knowledge, outbound resources aren't usually written to after their defaults are set.

nklaassen replied:
thanks for the review. I agree, I've updated this in the latest commit

nklaassen committed:
mention v4 roles in docs

nklaassen committed:
docs update

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
copy role when downgrading instead of mutating

nklaassen committed:
lint fix

inertial-frame reviewed (approved):
Docs approved - thanks!

klizhentas reviewed (approved):
The logic and UX looks good. How would our existing preset roles behave with this change, in particular, "access" preset. Would it continue working?

nklaassen commented:
> The logic and UX looks good. How would our existing preset roles behave with this change, in particular, "access" preset. Would it continue working?

@klizhentas Yes, existing built-in and preset roles are remaining on V3 for now, there should be no noticeable change. In https://github.com/gravitational/teleport/pull/7132 all of the go types are changing to `RoleV4`, but the version tag and thus the defaults are staying on V3, this change should be completely invisible to users.

The "access" preset explicitly sets all labels to wildcard-allow anyways, and the "editor" preset does not set any label values and appears to rely on the V3 wildcard defaults to be set.

The only roles that should be affected here are ones explicitly created by users on V4. We may want to migrate built-in and preset roles (and possibly existing user roles) in a future major release.

nklaassen committed:
update min supported version to 6.2.4

nklaassen committed:
Merge branch 'master' into nklaassen/roles-v4

nklaassen committed:
Rename RoleV3 to RoleV4 (#7132)

nklaassen committed:
cosmetic edits
